docs/architecture/adr-055-orm.md

```diff
@@ -0,0 +1,105 @@
+and subject to change, it is hard for clients to generically index, query and verify merkle proofs against state data.
+
+The known first instance of an "ORM" in the Cosmos ecosystem was in [weave](https://github.com/iov-one/weave/tree/master/orm).
+A later version was built for [regen-ledger](https://github.com/regen-network/regen-ledger/tree/master/orm) for
+use in the group module and later [ported to the SDK](https://github.com/cosmos/cosmos-sdk/tree/main/x/group/internal/orm)
+just for that purpose.
+
+While these earlier designs made it significantly easier to write state machines, they still required a lot of manual
+configuration, didn't expose state format directly to clients, and were limited in their support of different types
+of index keys, composite keys, and range queries.
+
+Discussions about the design continued in https://github.com/cosmos/cosmos-sdk/discussions/9156 and more
+sophisticated proofs of concept were created in https://github.com/allinbits/cosmos-sdk-poc/tree/master/runtime/orm
+and https://github.com/cosmos/cosmos-sdk/pull/10454.
+
+## Decision
+
+These prior efforts culminated in the creation of the Cosmos SDK `orm` go module which uses protobuf annotations
+for specifying ORM table definitions. This ORM is based on the new `google.golang.org/protobuf/reflect/protoreflect`
+API and supports:
+* sorted indexes for all simple protobuf types (except `bytes`, `enum`, `float`, `double`) as well as `Timestamp` and `Duration`
+* unsorted `bytes` and `enum` indexes
+* composite primary and secondary keys
+* unique indexes
+* auto-incrementing `uint64` primary keys
+* complex prefix and range queries
+* paginated queries
+* complete logical decoding of KV-store data
+
+Almost all the information needed to decode state directly is specified in .proto files. Each table definition specifies
+an ID which is unique per .proto file and each index within a table is unique within that table. Clients then only need
+to know the name of a module and the prefix ORM data for a specific .proto file within that module in order to decode
+state data directly. This additional information will be exposed directly through app configs which will be explained
+in a future ADR related to app wiring.
+
+The ORM makes optimizations around storage space by not repeating values in the primary key in the key value
+when storing primary key records. For example, if the object `{"a":0,"b":1}` has the primary key `a`, it will
+be stored in the key value store as `Key: '0', Value: {"b":1}` (with more efficient protobuf binary encoding).
+Also, the generated code from https://github.com/cosmos/cosmos-proto does optimizations around the
+`google.golang.org/protobuf/reflect/protoreflect` API to improve performance.
+
+A code generator is included with the ORM which creates type safe wrappers around the ORM's dynamic `Table`
+implementation and is the recommended way for modules to use the ORM.
```

Comment:
I personally think a subsection with an example would be useful to show how the ORM concretely looks like for modules:
- the proto options for a table
- generated code (at least the generated methods signatures)
- how it's used in a keeper

Probably links to the test bank ORM table would suffice